tests/instruments/test_user_api.py

```diff
@@ -0,0 +1,125 @@
+            if_type = self.adapter.connection.resource_info.interface_type
+            # Different communication methods might have different valid and/or needed attributes
+            # This is just an example on how this is possible since pyvisa includes
+            # several type of interfaces
+            if if_type is InterfaceType.asrl and not 'baud_rate' in kwargs:
+                self.adapter.connection.baud_rate = 2400
+            elif if_type is InterfaceType.gpib and not 'enable_repeat_addressing' in kwargs:
+                self.adapter.connection.enable_repeat_addressing = False
+
+# The defined tests use the pyvisa-sim simulated connections to avoid the need for a connected
+# instrument. The argument handling and connection creation happens in pyvisa just like for a
+# real instrument, through.
+
+
+def test_serial_default_settings():
+    """As a user, I want to simply connect to an instrument using default settings"""
+    instr = MultiprotocolInstrument(adapter='ASRL1::INSTR', visa_library='@sim')
+    assert instr.adapter.connection.baud_rate == 2400
+
+
+def test_serial_custom_baud_rate_is_set():
+    """As a user, I want to easily override default settings to fit my needs"""
+    instr = MultiprotocolInstrument(adapter='ASRL1::INSTR', baud_rate=115200, visa_library='@sim')
+    assert instr.adapter.connection.baud_rate == 115200
+
+
+def test_connections_use_tcpip():
+    """As a user, I want to be free to choose which connection to use (e.g. serial-over RS-232,
+    USB, TCP/IP) should an instrument support more than one
+    """
+    # here, baud_rate is invalid for pyvisa, but is a default kwarg anyway
+    instr = MultiprotocolInstrument(adapter='TCPIP::localhost:1111::INSTR', visa_library='@sim')
+    assert hasattr(instr.adapter.connection, 'control_ren') # method specific to pyvisa TCPIPInstrument
+
+
+def test_connections_use_gpib():
+    """As a user, I want to be free to choose which connection to use (e.g. serial-over RS-232,
+    USB, TCP/IP) should an instrument support more than one
+    """
+    # here, baud_rate is invalid for pyvisa, but is a default kwarg anyway
+    instr = MultiprotocolInstrument(adapter='GPIB::8::INSTR', visa_library='@sim')
+    assert instr.adapter.connection.enable_repeat_addressing == False  # attribute specific to pyvisa GPIBInstrument
+
+
+def test_use_separate_adapter_instance():
+    """As a user, I want to be able to supply a self-generated Adapter instance
+    (e.g. to enable serial connection sharing over RS-485/-422)
+    """
+    ser = SerialAdapter(serial.serial_for_url("loop://", baudrate = 4800)) # note: not baud_rate, as this goes to pyserial
+    instr = MultiprotocolInstrument(ser)  # don't need visa_library here as this does not go to pyvisa, kwargs are ignored if provided
+    print (ser, ser.connection)
+    print (instr.adapter)
+    assert instr.adapter.connection.baudrate == 4800
+
+def test_use_separate_adapter_instance1():
```

Comment:
Maybe better name for these tests would be `test_use_separate_SerialAdapter` and `test_use_separate_VISAAdapter`

Reply:
Indeed